include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl

```diff
@@ -3,428 +3,573 @@
+		const int32_t lowLsb = find_lsb_helper<uint32_t>::__call(lowBits);
+
+		if (lowLsb == -1)
+		{
+			const uint32_t highBits = uint32_t(val >> 32);
+			const int32_t highLsb = find_lsb_helper<uint32_t>::__call(highBits);
+			if (highLsb == -1)
+				return -1;
+			else
+				return 32 + highLsb;
+		}
+
+		return lowLsb;
+	}
+};
+
+#define AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER_2_ARG_FUNC(HELPER_NAME, SPIRV_FUNCTION_NAME, RETURN_TYPE)\
+template<typename T> NBL_PARTIAL_REQ_TOP(always_true<decltype(spirv::SPIRV_FUNCTION_NAME<T>(experimental::declval<T>()))>)\
+struct HELPER_NAME<T NBL_PARTIAL_REQ_BOT(always_true<decltype(spirv::SPIRV_FUNCTION_NAME<T>(experimental::declval<T>()))>) >\
+{\
+	using return_t = RETURN_TYPE;\
+	static inline return_t __call(const T a, const T b)\
+	{\
+		return spirv::SPIRV_FUNCTION_NAME<T>(a, b);\
+	}\
+};
 
-template<int N>
-struct find_msb_helper<vector<uint32_t, N> >
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER_2_ARG_FUNC(max_helper, fMax, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER_2_ARG_FUNC(max_helper, uMax, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER_2_ARG_FUNC(max_helper, sMax, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER_2_ARG_FUNC(min_helper, fMax, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER_2_ARG_FUNC(min_helper, uMax, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER_2_ARG_FUNC(min_helper, sMax, T)
+
+#else // C++ only specializations
+
+template<typename T>
+requires concepts::Scalar<T>
+struct clamp_helper<T, T>
 {
-	static vector<int32_t, N> findMSB(NBL_CONST_REF_ARG(vector<uint32_t, N>) val)
+	using return_t = T;
+	static inline return_t __call(const T val, const T min, const T max)
 	{
-#ifdef __HLSL_VERSION
-		return spirv::findUMsb(val);
-#else
-		return glm::findMSB(val);
-#endif
+		return std::clamp(val, min, max);
+	}
+};
+template<typename T>
+requires concepts::IntegralScalar<T>
+struct bitReverse_helper<T>
+{
+	static inline T __call(NBL_CONST_REF_ARG(T) arg)
+	{
+		return glm::bitfieldReverse(arg);
 	}
 };
+template<typename Matrix>
+requires concepts::Matrix<Matrix>
+struct transpose_helper<Matrix>
+{
+	using transposed_t = typename matrix_traits<Matrix>::transposed_type;
 
-template<int N>
-struct find_msb_helper<vector<int32_t, N> >
+	static transposed_t __call(NBL_CONST_REF_ARG(Matrix) m)
+	{
+		return reinterpret_cast<transposed_t&>(glm::transpose(reinterpret_cast<typename Matrix::Base const&>(m)));
+	}
+};
+template<typename Vector>
+requires concepts::FloatingPointVector<Vector>
+struct length_helper<Vector>
 {
-	static vector<int32_t, N> findMSB(NBL_CONST_REF_ARG(vector<int32_t, N>) val)
+	static inline typename vector_traits<Vector>::scalar_type __call(NBL_CONST_REF_ARG(Vector) vec)
+	{
+		return std::sqrt(dot_helper<Vector>::__call(vec, vec));
+	}
+};
+template<typename Vectorial>
+requires concepts::FloatingPointLikeVectorial<Vectorial>
+struct normalize_helper<Vectorial>
+{
+	static inline Vectorial __call(NBL_CONST_REF_ARG(Vectorial) vec)
+	{
+		return vec / length_helper<Vectorial>::__call(vec);
+	}
+};
```

Comment:
shouldn't this be common to C++ and HLSL at the bottom, one can't call SPIR-V intrinsics on this

Reply:
this is c++ only specialization, and hlsl only specialization is [here](https://github.com/Devsh-Graphics-Programming/Nabla/blob/new_tgmath/include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl#L115), it works fine